File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> fre;
        vector<int> res;
        for(int i = 0; i < nums.size(); i++){
            fre[nums[i]]++;
        }
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minheap;
        for(auto i : fre){
            minheap.push({i.second, i.first});
            if(minheap.size() > k){
                minheap.pop();
            }
        }
        while(!minheap.empty()){
            res.push_back(minheap.top().second);
            minheap.pop();
        }
        return res;
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> freqMap = new HashMap<>();
        for(int num : nums){
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }

        PriorityQueue<int[]> heap = new PriorityQueue<>(
            (a, b) -> a[1] - b[1]
        );
        for(Map.Entry<Integer, Integer> entry : freqMap.entrySet()){
            heap.offer(new int[]{entry.getKey(), entry.getValue()});
            if(heap.size() > k){
                heap.poll();
            }
        }
        int[] res = new int[k];
        for(int i = k - 1; i >= 0; i--){
            res[i] = heap.poll()[0];
        }
        return res;
    }
}